File: MatlabCodes/Fig06_02.m
function COS_LogNormal_Density_Recovery
close all;clc;
format long
i = complex(0,1); %assigning i=sqrt(-1)

% COS method settings

a = -10; % Truncation domain
b =  10;

% Characteristic function parameters

mu     = 0.5;    sigma  = 0.2;

% Domain for the density f(x)

y = linspace(0.05,5,1000);

% Characteristic function  

cf  =  @(u) exp(i * mu * u - 0.5 * sigma^2 * u.^2);

% Iterate over different numbers of expansion terms

Ngrid = [16,64,128];
idx = 1;
figure(1); hold on; grid on;xlabel('y');ylabel('f_Y(y)')
legend_txt= cell(length(Ngrid),1);

for N=Ngrid

    % Density from the COS method

    f_Y =1./y .* COSDensity(cf,log(y),N,a,b);
    plot(y,f_Y)
    legend_txt{idx} = sprintf('N=%i',N);
    idx = idx + 1;
end
legend(legend_txt)

function f_X = COSDensity(cf,x,N,a,b)
i = complex(0,1); %assigning i=sqrt(-1)
k = 0:N-1; 
u = k * pi / (b - a);

% F_k coefficients

F_k    = 2 / (b - a) * real(cf(u) .* exp(-i * u * a));
F_k(1) = F_k(1) * 0.5; % adjustment for the first term

% Final calculation

f_X = F_k * cos(u' * (x - a));

File: PythonCodes/Chapter 06/Fig06_02.py
#%%
"""
LogNormal density recovery using the COS method
"""
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as st

def COSDensity(cf,x,N,a,b):
    i = np.complex(0.0,1.0) #assigning i=sqrt(-1)
    k = np.linspace(0,N-1,N)
    u = np.zeros([1,N])
    u = k * np.pi / (b-a)
        
    # F_k coefficients

    F_k    = 2.0 / (b - a) * np.real(cf(u) * np.exp(-i * u * a));
    F_k[0] = F_k[0] * 0.5; # adjustment for the first term
    
    # Final calculation

    f_X = np.matmul(F_k , np.cos(np.outer(u, x - a )))
        
    # We output only the first row

    return f_X
    
def mainCalculation():
    i = np.complex(0.0, 1.0) #assigning i=sqrt(-1)
    
    # Setting for the COS method 

    a = -10
    b = 10
    
    # Define the range for the expansion terms

    N = [16, 64, 128]
    
    # Parameter setting for normal distribution

    mu = 0.5
    sigma = 0.2 
        
    # Define characteristic function for the normal distribution

    cF = lambda u : np.exp(i * mu * u - 0.5 * np.power(sigma,2.0) * np.power(u,2.0));
    
    # Define the domain for density

    y = np.linspace(0.05,5,1000)
        
    plt.figure(1)
    plt.grid()
    plt.xlabel("y")
    plt.ylabel("$f_Y(y)$")
    for n in N:
        f_Y = 1/y * COSDensity(cF,np.log(y),n,a,b)
        plt.plot(y,f_Y)
    plt.legend(["N=%.0f"%N[0],"N=%.0f"%N[1],"N=%.0f"%N[2]])
    
    
mainCalculation()
